src/engine.ml

```diff
@@ -183,6 +183,52 @@ let client config ts now rng =
         let remotes = Array.to_list remotes in
         Config.add Remote remotes config
   in
+  let* config =
+    match Config.find Pkcs12 config with
+    | None -> Ok config
+    | Some p12 -> (
+        let* pass =
+          Option.to_result ~none:(`Msg "missing pkcs12 password")
+            pkcs12_password
+        in
+        let* stuff = X509.PKCS12.verify pass p12 in
+        let cert =
+          List.filter_map
+            (function `Certificate c -> Some c | _ -> None)
+            stuff
+        and key =
+          List.filter
+            (function
+              | `Private_key _ | `Decrypted_private_key _ -> true | _ -> false)
+            stuff
+        in
+        match (cert, key) with
+        | certs, ([ `Private_key pk ] | [ `Decrypted_private_key pk ]) -> (
+            let key_fp =
+              X509.Public_key.fingerprint (X509.Private_key.public pk)
+            in
+            match
+              List.filter
+                (fun cert ->
+                  let cert_pubkey = X509.Certificate.public_key cert in
+                  Cstruct.equal (X509.Public_key.fingerprint cert_pubkey) key_fp)
+                certs
+            with
+            | [ cert ] ->
+                Ok
+                  Config.(
+                    add Tls_key pk (add Tls_cert cert (remove Pkcs12 config)))
+            | _ ->
```

Comment:
Apparently `--pkcs12` can also be used instead of `--ca`. So I wonder if `--pkcs12` and `--ca` are mutually exclusive. I will investigate.
```suggestion
            match
              List.partition
                (fun cert ->
                  let cert_pubkey = X509.Certificate.public_key cert in
                  Cstruct.equal (X509.Public_key.fingerprint cert_pubkey) key_fp)
                certs
            with
            | [ cert ], _ :: _ as cas ->
                Ok
                  Config.(
                    add Ca cas (add Tls_key pk (add Tls_cert cert (remove Pkcs12 config))))
            | _ ->
```

Reply:
hmm, but there may be as well a certificate chain in the PKCS12 (to be send to the server)... I'm not entirely sure how OpenVPN handles this tbh.

also, the CA for the server may be different than the CA for the user authentication (i.e. a server may have a public dns name, and use let's encrypt certs -- while the client gets a certificate from a private CA).

Reply:
It seems `--pkcs12` and `--ca` are not mutually exclusive https://github.com/OpenVPN/openvpn/blob/a1cb1b47b138b9f654cd0bca5de6d08dbca61888/src/openvpn/options.c#L2952-L2978 even if `--pkcs12` and `--capath` are mutually exclusive (odd!). I am not sure what the semantics are.

Reply:
Yes, this is very confusing. There is as well `--extra-certs` which are certificates in the chain of the local certificate and not used for authenticating the remote (while `--ca` can be used for both). It would make more sense if `--pkcs12` can be used instead of `--extra-certs` (and not `--ca`).